### PR TITLE
Using cmake as build system for Windows, Mac, and Linux

### PR DESCRIPTION
**Pull request by scottenglert**
I have added a cmake build system that will make setting up the dependencies and work across Windows, Mac, and Linux (untested). Mac was a difficult problem as Maya's devkit does not ship with the Python headers. So to not require the developer to have all the correct Python versions installed or have Maya installed, I copied the correct headers from the Maya installation including the shared libraries.

I have also added Pybind11 as a submodule so it will be easier to manage it rather than just a copy in the repository. It has to be version 2.9.0 as that is the last version to support Python 2.

I have successfully build Maya 2020 and 2022 on both Windows and Mac. I haven't tested it in linux yet. Some tweaks can still be made, but it works fine for now.

### CODE REVIEW

**Comment by mottosso** [comment]
Thanks @scottenglert.

This one is going to take some convincing, because there already is a build system setup to run automatically on each PR and especially each GitHub release made.

- https://github.com/mottosso/cmdc/blob/master/.github/workflows/main.yml

That automatically produces releases with these goods.

image

As setting this up took quite a bit of sweat, blood and tears I'm not terribly keen on revisiting this. 😅 Would you be up for replacing this mechanic with your CMake system so we can make some comparisons? I'd also be keen to hear what it is you find easier to manage with CMake than the plain shell scripts currently in place.

Also, it appears the build environment has changed since last built; possibly a newer version of Python got installed.

**Comment by scottenglert** [comment]
Thanks for the feedback, I am not too savvy with Github's workflows, so I need to learn more about them. I understand what they do in concept and didn't know they actually produced a download.  Do you know how this works with a Mac environment? Have to read more about what is possible with github.

My main point on cmake for me is setting up all the compiler flags, choosing the generated build system, and having things in one unit. It's my preference to mange all the builds on operating systems and maya versions, so I can keep this on my fork if you like. I can take a stab at porting the github workflows to use cmake and see how hard it is like you said. Cmake is more of comfort level than writing bash or powershell scripts.

**Comment by mottosso** [comment]
> Do you know how this works with a Mac environment? 

There's a comment-out section for Mac towards the bottom of that configuration file, it should provide a vanilla Mac environment with CMake etc. I think the problem I was having was what you ran into as well, about missing Python sources in the Maya devkit.

**Comment by scottenglert** [comment]
Yeah, I am trying to get the gitlab runner working, I almost have the windows one, just have to work through a couple differences in the the windows server they provide.

**Comment by mottosso** [comment]
Just want to chime in to say you're doing the right thing; this workflow of pushing a config to see how GitHub will react is the only way I know how to do it too. It's painful and slow, but that's the world we live in. 🤷‍♂️

**Comment by scottenglert** [comment]
Yeah, I think I am almost done, I see they added the 2023 devkit to download, so I am going to add them as targets, but it is using Python 3.9.

Sorry if you see a lot of spam emails about failed jobs. I'll report back once they are all building.

**Comment by scottenglert** [comment]
I think I have all the build working now, but I have a problem with the upload-artifacts for some reason not finding the cmdc.so in the build directory in the linux builds. I was trying to read if this is a known issue as it appears this don't working containers.

I looked at your runner and I have the same thing, I have event printed out all the files in the build directory and see that it's there, but the upload-artifact isn't finding it. Any thoughts?

**Comment by scottenglert** [comment]
Nevermind, I figured it out, the build was saved on the runner but the artifact was looking in the container, I changed it so the build was stored on the container and it works. Confusing between the runner and container directories with the workspaces. Learning as I go! But this completes the build. I am working on the stubs part next.

**Comment by scottenglert** [comment]
I think I got all the builds and stubs working now, let me know what you think. I made my own docker images as they are based off the of the Academy Software Foundation images that are built for the VFX platform so our compiler and libraries are matching what Maya recommends for compatibility.  So yeah, I haven't tested the final zip archive but all the pieces are there and should be easy to add new Maya versions and Python.

**Comment by mottosso** [comment]
Hey @scottenglert, thanks for suffering through this; I think at this point you know the pain I felt when setting up the first iteration of this. 😅 Can be fun though! Good to have these boring bits out of the way and automated.

I gave your repo a quick go:

```bash
cmake -B ./build -DCMAKE_BUILD_TYPE=Release -DMAYA_VERSION=2020 -DMAYA_DEVKIT_ROOT="c:\mayadevkit\2020"
cmake --build .\build --config Release
# Microsoft (R) Build Engine version 16.11.0+0538acc04 for .NET Framework
# Copyright (C) Microsoft Corporation. All rights reserved.
# 
#   Checking Build System
#   Parsing MFn.h to MFn.Types.inl...
#   Building Custom Rule C:/github/scottenglert/cmdc/CMakeLists.txt
#   Building Custom Rule C:/github/scottenglert/cmdc/CMakeLists.txt
#   main.cpp
#      Creating library C:/github/scottenglert/cmdc/build/Release/cmdc.lib and object C:/github/scottenglert/cmdc/build/Release/cmdc.
#   exp
#   cmdc.vcxproj -> C:\github\scottenglert\cmdc\build\Release\cmdc.pyd
#   Clearing MFn.Types.inl...
#   Building Custom Rule C:/github/scottenglert/cmdc/CMakeLists.txt
```

And just to be sure, I gave it a quick run.

```py
>>> import cmdc
>>> dir(cmdc)
['Angle', 'BoundingBox', 'Color', 'DAGDrawOverrideInfo', 'DGModifier', 'DagModifier', 'DagPath', 'DagPathArray', 'DataBlock', 'DataHandle', 'Distance', 'Fn', 'FnDagNode', 'FnDependencyNode', 'Matrix', 'NodeClass', 'Object', 'ObjectArray', 'ObjectHandle', 'Plug', 'Point', 'PxData', 'Quaternion', 'SelectionList', 'Status', 'String', 'Time', 'TypeId', 'Uuid', 'Vector', '__doc__', '__file__', '__name__', '__package__', '__version__']
>>> mod = cmdc.DagModifier()
>>> t = mod.createNode("transform")
>>> mod.doIt()
>>> t
<cmdc.Object(kTransform)>
>>> dir(t)
['__class__', '__delattr__', '__doc__', '__eq__', '__format__', '__getattribute__', '__hash__', '__init__', '__module__', '__new__', '__qualname__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'apiType', 'apiTypeStr', 'hasFn', 'isNull', 'kNullObj']
```

And it worked wonderfully. 😃 

A few concerns though.

1. We lost Maya 2018 and 2019 in the process.. People still use these. We did gain 2023, so that's a plus.
2. What was wrong with the container we were already using? 🤔 It's one I've used for many years, until and including Maya 2022, for both open source and commercially released software that is being tested thoroughly on a daily basis.
3. Is there a Dockerfile for your container? Best I can find is [this](https://registry.hub.docker.com/layers/maya-build/scottenglert/maya-build/2022.3/images/sha256-8f4e99f25c39aeab2188543b60b94b9d25110cf619dc3f99e45a653cfe133f1a?context=explore). The one currently in use [is here](https://github.com/mottosso/docker-maya)
4. The Python libraries have [grown in size](https://github.com/scottenglert/cmdc/tree/master/os); as mentioned in the README there, the libraries are identical across versions of Maya, so it's kind of wasteful having copies of these. But if we keep them, we should probably update the README also.
5. That's all I could find!

File: Dockerfile.Base [hunk 1]
```diff
@@ -0,0 +1,17 @@
+FROM gcc:6.3
```
Comment (mottosso) [inline Dockerfile.Base:1]: Is this Dockerfile in use? Can't find where, maybe it was a temporary one?

**Comment by scottenglert** [comment]
Yeah it was a bit of a pain with lack of clear examples and docs, I copied a lot what you did which was a huge help.

So to answer:

1. We can add that those back, it was mostly because I can't find or have the installations for those to run mayapy for the stub generation. The down side to having so many supported version is the api changes and trying to manage those, we might have to possibly have a different '.inl' file depending on which version of Maya. If you have a location to download Maya 2018 and 2019 for linux, I can put that in my containers. Which leads to...
2. When I pulled your containers, I noticed the versions of gcc don't match up with what Maya recommends and the vfx platform (which should match). For example your 2020.1 image is using gcc 4.8. 5 unless it's installed somewhere else? Now there shouldn't be that much with compatibility problems, but to be able to use c++17 for example it would be nice to follow the newer allowed compilers. That is all, I have run into very specific crashes due to different gcc versions with Maya's API years a go, also hopefully newer compiler means better optimizations and other supports in newer c++ features. I based my images with the Academy Software Foundation docker contains that have this already setup so it doesn't require any more than just downloading and extracting the dev kit.
3. Sure, I could put the docker files in the repo if other people want to use them or add a link. I didn't know who you set up your containers so I just pushed one of my local ones I have used.
4. I am not sure what you mean by keep them, keep what exactly? There is one version of each Python version (2.7, 3.7, 3.9) per OS. You need the OS specific lib to link to. Sorry, bit lost there?

So if you like, I can add back in the Maya 2018 and 2019 now that I think I can build the right containers for them. Again, I'm not trying to reinvent the wheel on everything, I am just trying to target everything as accurately as possible and make it so it would be easier for a studio for example to pull and build the source easily, I am thinking about doing this at my studio if the API has more coverage.

Thanks again for your feedback and taking the first hit setting up all the scripts and actions.

**Comment by mottosso** [comment]
> If you have a location to download Maya 2018 and 2019 for linux, I can put that in my containers.

Yes, see here.

- https://github.com/mottosso/docker-maya/blob/2018.7/Dockerfile

> I based my images with the Academy Software Foundation docker contains that have this already setup so it doesn't require any more than just downloading and extracting the dev kit.

Ok, in that case it would be good to try and compare to see what is gained/lost. Could you include relevant links? And your Dockerfile(s) too would be great; if it isn't public elsewhere then it'd be fine including them in this repo.

> Sure, I could put the docker files in the repo if other people want to use them or add a link. I didn't know who you set up your containers so I just pushed one of my local ones I have used.

For reproducibility's sake, we should definitely have the source somewhere. At the moment, it's unclear when another update is made that breaks the build for cmdc, making it that much harder to track down why.

> I am not sure what you mean by keep them, keep what exactly? There is one version of each Python version (2.7, 3.7, 3.9) per OS. You need the OS specific lib to link to. Sorry, bit lost there?

Sorry! That's right, I saw this.

**Before**



**After**



And figured each OS also had a version for each version of Maya; but it just so happens that these three versions have their own unique version of Python. Before, 2018-2020 had the same Python lib.

There is one thing we can safely remove however, I think? Which is the Python `include/` directory. I think this is actually included in the Maya devkit, and would slim down this repo by about 500 files.

**Comment by scottenglert** [comment]
I have the 2018 and 2019 builds added. I had to include the Python "include" directory on the mac os because funny enough it's not inside the dev kit, only in a specific directory on the install, a bit more of a mac framework directory structure, which is why I copied it to the repo so not to require the install just to build.

Sure, I'll move my docker files into the repo. The only thing is 2018 maya uses gcc 4.8 and that predates any Academy Software Foundation images, so I had to reuse the 2019 one which uses gcc 6.3. Maybe one of your containers for 2018 would work instead?

I'll also clean up some of the directories if we do go with cmake, then the shell scripts are not needed. I'll post more info about the docker images but for now: [https://hub.docker.com/r/aswf/ci-base](https://hub.docker.com/r/aswf/ci-base)

**Comment by mottosso** [comment]
> Maybe one of your containers for 2018 would work instead?

Hm. Considering Autodesk just removed 2018 from their downloads in favour of 2023, maybe it's safe to exclude 2018 from here if it complicates things.

One more thing I just realised; we lost the test suite! That would help narrow confirm each build actually works. Here's where that happens in the current GitHub workflow.

https://github.com/mottosso/cmdc/blob/master/.github/workflows/main.yml#L148

I expect we could reuse (copy/paste) most if not all of that, since it is independent from the build steps. Ideally we'd run tests on all platforms, but I haven't got a Docker container for Windows and Maya, nor MacOS. It has in my experience been exceedingly rare that something interacting with the Maya API which works on one platform does not work on the other, so it should be sufficient.

**Comment by scottenglert** [comment]
I have cleaned up the repo to remove the original scripts in my branch, copied over the docker files into the docker folder. Got the 2023 test added.

All the tests runs, but they all fail with:

```
Check that the Fn.Type enum is identical to the MFn.Type enum ... FAIL

======================================================================
FAIL: Check that the Fn.Type enum is identical to the MFn.Type enum
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/github/home/.local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/__w/cmdc/cmdc/tests/test_mfn.py", line 15, in test_fn_types
    "Missing enum '%s'." % k)
AssertionError: Missing enum 'kData2Long'.
```

**Comment by scottenglert** [comment]
Sorry, fixed one thing and its just 2018 and 2019 that fail the unit tests. Probably those Maya versions do not have those types, but 2020+ do. Have to see what the cause is...

**Comment by scottenglert** [comment]
Found that in the header, there are 4 legacy / backward compatibility types that are no longer used. The test was picking them up, but the MFn.inl didn't have them because they appear after kLast. So I have just added to the test to ignore these types as they also do not appear in Maya's Python 2.0 API.

**Comment by mottosso** [comment]
Fantastic! Ok, let's give this a whirl. :) Great work @scottenglert

**Comment by mottosso** [comment]
Looks like we've got a success!

- https://github.com/mottosso/cmdc/releases/tag/v0.1.8

**Pull request merged**